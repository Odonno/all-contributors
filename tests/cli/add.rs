use assert_fs::TempDir;
use color_eyre::eyre::{Error, Result};
use insta::{Settings, assert_snapshot};
use std::fs;

use crate::helpers::{
    InstaSettingsExtensions, copy_config_file, copy_readme_file, create_cmd, get_stderr_str,
    get_stdout_str,
};

#[test]
fn fails_to_add_if_no_config_file() -> Result<()> {
    let temp_dir = TempDir::new()?;

    let mut cmd = create_cmd(&temp_dir)?;

    cmd.env("NO_COLOR", "1").arg("add");

    let assert = cmd.assert().try_failure()?;
    let stderr = get_stderr_str(assert)?;

    let mut insta_settings = Settings::new();
    insta_settings.add_cli_location_filter();
    insta_settings.bind(|| {
        assert_snapshot!(stderr);
        Ok::<(), Error>(())
    })?;

    temp_dir.close()?;

    Ok(())
}

#[test]
fn add_an_existing_contribution_should_do_nothing() -> Result<()> {
    let temp_dir = TempDir::new()?;

    copy_config_file(&temp_dir)?;

    let mut cmd = create_cmd(&temp_dir)?;

    cmd.env("NO_COLOR", "1")
        .arg("add")
        .arg("tekacs")
        .arg("code");

    let assert = cmd.assert().try_success()?;
    let stdout = get_stdout_str(assert)?;

    let updated_config_file = fs::read_to_string(temp_dir.join(".all-contributorsrc"))?;

    let mut insta_settings = Settings::new();
    insta_settings.add_cli_location_filter();
    insta_settings.bind(|| {
        assert_snapshot!(stdout);
        assert_snapshot!(updated_config_file);
        Ok::<(), Error>(())
    })?;

    temp_dir.close()?;

    Ok(())
}

#[test]
fn should_add_new_contribution_to_existing_contributor() -> Result<()> {
    let temp_dir = TempDir::new()?;

    copy_config_file(&temp_dir)?;

    let mut cmd = create_cmd(&temp_dir)?;

    cmd.env("NO_COLOR", "1")
        .arg("add")
        .arg("tekacs")
        .arg("ideas,bug");

    let assert = cmd.assert().try_success()?;
    let stdout = get_stdout_str(assert)?;

    let updated_config_file = fs::read_to_string(temp_dir.join(".all-contributorsrc"))?;

    let mut insta_settings = Settings::new();
    insta_settings.add_cli_location_filter();
    insta_settings.bind(|| {
        assert_snapshot!(stdout);
        assert_snapshot!(updated_config_file);
        Ok::<(), Error>(())
    })?;

    temp_dir.close()?;

    Ok(())
}

#[test]
fn should_add_new_contributor_and_its_contributions() -> Result<()> {
    let temp_dir = TempDir::new()?;

    copy_config_file(&temp_dir)?;

    let mut cmd = create_cmd(&temp_dir)?;

    cmd.env("NO_COLOR", "1")
        .arg("add")
        .arg("odonno")
        .arg("code,ideas");

    let assert = cmd.assert().try_success()?;
    let stdout = get_stdout_str(assert)?;

    let updated_config_file = fs::read_to_string(temp_dir.join(".all-contributorsrc"))?;

    let mut insta_settings = Settings::new();
    insta_settings.add_cli_location_filter();
    insta_settings.bind(|| {
        assert_snapshot!(stdout);
        assert_snapshot!(updated_config_file);
        Ok::<(), Error>(())
    })?;

    temp_dir.close()?;

    Ok(())
}

#[test]
fn should_add_new_contributor_and_its_contributions_and_then_apply_generate() -> Result<()> {
    let temp_dir = TempDir::new()?;

    copy_config_file(&temp_dir)?;
    copy_readme_file(&temp_dir)?;

    let mut cmd = create_cmd(&temp_dir)?;

    cmd.env("NO_COLOR", "1")
        .arg("add")
        .arg("odonno")
        .arg("code,ideas")
        .arg("--generate");

    let assert = cmd.assert().try_success()?;
    let stdout = get_stdout_str(assert)?;

    let updated_config_file = fs::read_to_string(temp_dir.join(".all-contributorsrc"))?;
    let updated_readme = fs::read_to_string(temp_dir.join("readme.md"))?;

    let mut insta_settings = Settings::new();
    insta_settings.add_cli_location_filter();
    insta_settings.bind(|| {
        assert_snapshot!(stdout);
        assert_snapshot!(updated_config_file);
        assert_snapshot!(updated_readme);
        Ok::<(), Error>(())
    })?;

    temp_dir.close()?;

    Ok(())
}

#[test]
fn should_add_new_contributor_and_its_contributions_and_then_apply_generate_alt() -> Result<()> {
    let temp_dir = TempDir::new()?;

    copy_config_file(&temp_dir)?;
    copy_readme_file(&temp_dir)?;

    let mut cmd = create_cmd(&temp_dir)?;

    cmd.env("NO_COLOR", "1")
        .arg("add")
        .arg("odonno")
        .arg("code,ideas")
        .arg("--gen");

    let assert = cmd.assert().try_success()?;
    let stdout = get_stdout_str(assert)?;

    let updated_config_file = fs::read_to_string(temp_dir.join(".all-contributorsrc"))?;
    let updated_readme = fs::read_to_string(temp_dir.join("readme.md"))?;

    let mut insta_settings = Settings::new();
    insta_settings.add_cli_location_filter();
    insta_settings.bind(|| {
        assert_snapshot!(stdout);
        assert_snapshot!(updated_config_file);
        assert_snapshot!(updated_readme);
        Ok::<(), Error>(())
    })?;

    temp_dir.close()?;

    Ok(())
}
