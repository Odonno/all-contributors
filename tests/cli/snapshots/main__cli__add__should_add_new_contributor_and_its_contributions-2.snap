---
source: tests/cli/add.rs
expression: updated_config_file
---
{
  "commit": false,
  "commitConvention": "gitmoji",
  "contributors": [
    {
      "avatar_url": "https://avatars.githubusercontent.com/u/63247?v=4",
      "contributions": [
        "code"
      ],
      "login": "tekacs",
      "name": "Amar Sood",
      "profile": "http://www.tekacs.com/"
    },
    {
      "avatar_url": "https://avatars.githubusercontent.com/u/21016014?v=4",
      "contributions": [
        "code",
        "ideas",
        "infra"
      ],
      "login": "Roms1383",
      "name": "Rom's",
      "profile": "https://github.com/Roms1383"
    },
    {
      "avatar_url": "https://avatars.githubusercontent.com/u/10587083?v=4",
      "contributions": [
        "bug"
      ],
      "login": "pranaypratyush",
      "name": "Pranay Pratyush",
      "profile": "https://github.com/pranaypratyush"
    },
    {
      "avatar_url": "https://avatars.githubusercontent.com/u/5788284?v=4",
      "contributions": [
        "ideas"
      ],
      "login": "ansarizafar",
      "name": "Zafar Ansari",
      "profile": "https://github.com/ansarizafar"
    },
    {
      "avatar_url": "https://avatars.githubusercontent.com/u/28957605?v=4",
      "contributions": [
        "ideas"
      ],
      "login": "turulix",
      "name": "Tim",
      "profile": "https://turulix.de/"
    },
    {
      "avatar_url": "https://avatars.githubusercontent.com/u/3688989?v=4",
      "contributions": [
        "bug"
      ],
      "login": "snarkipus",
      "name": "Matt Jackson",
      "profile": "https://github.com/snarkipus"
    },
    {
      "avatar_url": "https://avatars.githubusercontent.com/u/53177682?v=4",
      "contributions": [
        "ideas"
      ],
      "login": "keyruu",
      "name": "Lucas",
      "profile": "https://github.com/keyruu"
    },
    {
      "avatar_url": "https://avatars.githubusercontent.com/u/6053067?v=4",
      "contributions": [
        "code",
        "ideas"
      ],
      "login": "odonno",
      "name": "David Bottiau",
      "profile": "https://dbottiau.fr"
    }
  ],
  "contributorsPerLine": 7,
  "files": [
    "readme.md"
  ],
  "imageSize": 100,
  "projectName": "surrealdb-migrations",
  "projectOwner": "Odonno",
  "repoHost": "https://github.com",
  "repoType": "github",
  "skipCi": true
}
